src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2000,11 +2008,11 @@ else if (!swept || e != prevRunState || (e & RS_LOCK) != 0) {
                     }
                     swept = (phaseSum == (phaseSum = sum));
                 }
-                else if (compareAndSetCtl(c, c) &&        // confirm
-                         casRunState(e, (e & SHUTDOWN) != 0 ? e | STOP : e)) {
-                    if ((e & SHUTDOWN) != 0)              // enable termination
-                        interruptAll();
+                else if ((e & SHUTDOWN) == 0)
                     return true;
+                else if (compareAndSetCtl(c, c) && casRunState(e, e | STOP)) {
```

Comment:
Is this `compareAndSetCtl(c, c)` really needed? Wouldn't a volatile get on `ctl` compared with `c` be enough? (Or are we relying on a store-barrier in the successful case? 🤔